the ScientificArtifact.py file was added, but not mixed in

dkeller9 opened the pull request:
This branch is just intended for evaluation by the scientific team.

james-isbister commented:
Several comments and questions:

IsDefinedByType, IsDefinedBy: 
- These terms aren't clear to me. Would something like "original_source_location" be more explicit?
- What would the value be for an scientific artifact generated by the platform? It is defined by the entity itself, no? An optional original_source_location parameter might be suitable?

subject_id:
- What would the subject be for an analysis of data across multiple species?
- What would the subject be for a new brain region model combining data from multiple "subjects"? 
- Maybe this would be better as an optional parameter?

contributions:
- What is the idea here?

license_id: 
- So the user has to select a license every time they create an artifact? I think this would slow users down, and is only needed once the entity is public

experiment_date
- Is there not already some storage if the creation date for entities made on the platform?

validation_tags:
- Indeed this might be a some type of entity itself. We might see how best to format validations when we implement the morphology validation in obi-one?

contact_email:
- Would this be part of the contact that is linked by the contact_id?

chr-pok commented:
Here are a few more comments/questions from my side:
- `brain_region_id`: What is that ID based on, some predefined list of regions? Couldn't it be a UUID as well referencing a unique region entity?
- Defined in [General_schema.docx](https://openbraininstitute.sharepoint.com/:w:/s/OBI-Scientificstaff/EQhs6O6doJ5GmEnNhVmeQCYBPLrBgs0njgUDSt0xVGam4w?e=4Bhngm) but seems to be missing here: `[optional] additional_brain_regions` in case more than one region is involved.
- I agree that`license_id` is only required when made public. Perhaps it could be optional.
- `experiment_date` should be optional as well. Not all scientific artifacts contain experimental data.
- I agree that `contact_email` should be better part of the entity that is referenced by `contact_id`. Also, it should be optional.
- `validation_tags`: Is it even possible to put a dict into the database? From earlier discussions I thought this would be handled by annotations which would contain more information than just a boolean value, like validation date, version of the asset being validated, version of the code used to run the validation, etc.

dkeller9 commented:
`IsDefinedByType, IsDefinedBy:

These terms aren't clear to me. Would something like "original_source_location" be more explicit?
What would the value be for an scientific artifact generated by the platform? It is defined by the entity itself, no? An optional original_source_location parameter might be suitable?`

I changed it to PublishedInType and added original_source_location parameter.

`subject_id:

What would the subject be for an analysis of data across multiple species?
What would the subject be for a new brain region model combining data from multiple "subjects"?
Maybe this would be better as an optional parameter?`

Now made optional.


`contributions:
What is the idea here?`

This is already in entity so removed.


`license_id:

So the user has to select a license every time they create an artifact? I think this would slow users down, and is only needed once the entity is public`

made optional

`experiment_date

Is there not already some storage if the creation date for entities made on the platform?`

Only creation and update date, not the actual experiment date.

`validation_tags:

Indeed this might be a some type of entity itself. We might see how best to format validations when we implement the morphology validation in obi-one?`

Sure, could be an entity. We had discussed adding tags in the science meeting.

`contact_email:

Would this be part of the contact that is linked by the contact_id?`

It is actually not in the contact class but maybe I will put it in there and remove it here.

`As there are some differences between entities created on the platform vs entities created externally, it might be worth considering having two types of mixin which inherit several shared properties?`

What other features would be needed that would justify the extra complexity?

dkeller9 commented:
> brain_region_id: What is that ID based on, some predefined list of regions? Couldn't it be a UUID as well referencing a unique region entity?

It's the id of the Brain Region entity.

> Defined in [General_schema.docx](https://openbraininstitute.sharepoint.com/:w:/s/OBI-Scientificstaff/EQhs6O6doJ5GmEnNhVmeQCYBPLrBgs0njgUDSt0xVGam4w?e=4Bhngm) but seems to be missing here: [optional] additional_brain_regions in case more than one region is involved.


Thanks, now added.

> I agree thatlicense_id is only required when made public. Perhaps it could be optional.
> experiment_date should be optional as well. Not all scientific artifacts contain experimental data.
> I agree that contact_email should be better part of the entity that is referenced by contact_id. Also, it should be optional.

These fields are now optional.

> validation_tags: Is it even possible to put a dict into the database? From earlier discussions I thought this would be handled by annotations which would contain more information than just a boolean value, like validation date, version of the asset being validated, version of the code used to run the validation, etc.

Further discussion is still needed.

chr-pok commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,38 @@
+class ScientificArtifactMixin(BaseModel):
+    name :str
+    description:str  
+    subject_id : uuid.UUID = None
```
If I am not mistaken, the correct notation should probably be:
```suggestion
    subject_id : uuid.UUID | None = None
```

chr-pok commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,38 @@
+    description:str  
+    subject_id : uuid.UUID = None
+    
+    brain_region_id: int
```
Just wondering why this ID is defined as `int`, whereas other IDs as `uuid.UUID`? Same applied to `additional_brain_regions`.

dkeller9 replied:
I speculate that it does not refer to the entity ID but rather the Id in the hierarchical atlas.

chr-pok commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,38 @@
+    subject_id : uuid.UUID = None
+    
+    brain_region_id: int
+    additional_brain_regions: list[int] | None
```
I would set it to None by default:
```suggestion
    additional_brain_regions: list[int] | None = None
```

chr-pok commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,38 @@
+    
+    brain_region_id: int
+    additional_brain_regions: list[int] | None
+    license_id: uuid.UUID = None #only needed when public 
```
Same as above:
```suggestion
    license_id: uuid.UUID | None = None #only needed when public 
```

chr-pok commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,38 @@
+    brain_region_id: int
+    additional_brain_regions: list[int] | None
+    license_id: uuid.UUID = None #only needed when public 
+    experiment_date: date = None 
```
Same as above:
```suggestion
    experiment_date: date | None = None 
```

chr-pok commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,38 @@
+    
+    validation_tags: dict[str, bool] #This is a dict{“properties_check”: T/F} (determined by a script not a user input. Should this be here or an annotation?) 
+
+    contact_id : uuid.UUID
```
I would say this should be optional as well:
```suggestion
    contact_id : uuid.UUID | None = None
```

chr-pok reviewed (commented):
Thanks for the changes so far! I have just added a few more comments/suggestions directly into the code.

jdcourcol reviewed (commented):
There is already a location mixin with brain_region_id.

jdcourcol commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,38 @@
+    original_source_location : str
+    other: str | None = None # Optional alternative identifier as a string
+
+class ScientificArtifactMixin(BaseModel):
```
why is this not derived from Entity ?

dkeller9 replied:
Not all entities will need things particular to scientific artifact, for example experiment date,

jdcourcol replied:
I agree. The question still stands though , why a ScientificArtifact does not derive from Entity ?

dkeller9 commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,38 @@
+    
+    validation_tags: dict[str, bool] #This is a dict{“properties_check”: T/F} (determined by a script not a user input. Should this be here or an annotation?) 
+
+    contact_id : uuid.UUID | None = None
```
This is the contact person if there are questions about the artifact.

jdcourcol replied:
I am confused, I don't see changes in the database model ?

dkeller9 commented:
> There is already a location mixin with brain_region_id.

Okay, brain region is removed and will come in through the mixin, although additional brain regions are retained. We might need to rethink how this is done. Also needed is a way to indicate which hierarchy classification schema the brain regions come from.

chr-pok commented:
> > There is already a location mixin with brain_region_id.
> 
> Okay, brain region is removed and will come in through the mixin, although additional brain regions are retained. We might need to rethink how this is done. Also needed is a way to indicate which hierarchy classification schema the brain regions come from.

Actually, I think in this case it would be nice to have additional brain regions directly added to the `BrainRegionFilterMixin` so it could be consistently used in other places as well. However, we should clearly specify what the purpose of `additional_brain_regions` actually is. Should it contain additional brain regions on the same level of granularity (e.g., brain region = S1FL, additional brain regions = [S1HL]), or a finer resolution of the (main) brain region (e.g., brain region = S1, additional brain regions = [S1FL, S1HL])?

dkeller9 commented:
> > > There is already a location mixin with brain_region_id.
> > 
> > 
> > Okay, brain region is removed and will come in through the mixin, although additional brain regions are retained. We might need to rethink how this is done. Also needed is a way to indicate which hierarchy classification schema the brain regions come from.
> 
> Actually, I think in this case it would be nice to have additional brain regions directly added to the `BrainRegionFilterMixin` so it could be consistently used in other places as well. However, we should clearly specify what the purpose of `additional_brain_regions` actually is. Should it contain additional brain regions on the same level of granularity (e.g., brain region = S1FL, additional brain regions = [S1HL]), or a finer resolution of the (main) brain region (e.g., brain region = S1, additional brain regions = [S1FL, S1HL])?

Okay, I have added  the additional brain regions directly to the `BrainRegionFilterMixin`. I see them as completely different brain regions rather than children of the main brain region. It is true though that the  BrainRegion already has a 'children' field already though. Perhaps this can be repurposed to be additonal_brain_regions.

jdcourcol commented on app/db/model.py:
```diff
@@ -143,47 +160,183 @@ class SpeciesMixin(Base):
+
+
+class ScientificArtifact(
+    Entity,
```
nice

jdcourcol commented on app/db/model.py:
```diff
@@ -143,47 +160,183 @@ class SpeciesMixin(Base):
+):
+    __tablename__ = "scientific_artifact"
+    id: Mapped[uuid.UUID] = mapped_column(ForeignKey("entity.id"), primary_key=True)
+    experiment_date: Mapped[datetime] = mapped_column(DateTime, nullable=True)
```
"experiment_date" : does that work well for a model ?

dkeller9 replied:
Yes, it should.

jdcourcol commented on app/db/model.py:
```diff
@@ -143,47 +160,183 @@ class SpeciesMixin(Base):
+    id: Mapped[uuid.UUID] = mapped_column(ForeignKey("entity.id"), primary_key=True)
+    experiment_date: Mapped[datetime] = mapped_column(DateTime, nullable=True)
+    published_in: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
+    validation_tags: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
```
I don't feel comfortable with a unstructured dict for this. This looks a recipe for mess.

dkeller9 replied:
For published_in or validation_tags or both?

jdcourcol replied:
for validation tags especially. I think this needs more specification. 
First, I don't think it should be a property of the scientific artefact but an assertion on the artefact. Which means that this is very close to the annotation pattern.
So more like another entity pointing to the artefact , with properties.
Second, who did run the validation, when , which code etc.. This should show up somewhere.

dkeller9 replied:
validation_tags is removed

jdcourcol commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,84 @@
+#        from_attributes = True  # Allow mapping from SQLAlchemy objects
+
+
+class ScientificArtifactCreate(BaseModel,BrainRegionFilterMixin):
```
should this derive from entityread?

dkeller9 replied:
Good point, I will investigate.

jdcourcol reviewed (changes requested):
I would say that we can merge the scientific artefact (assuming there are tests and end point added ..) as long as we remove the published_in & validation_tags for now.

I am against unstructured data for these use cases.
Having scientific artefact merged would help @chr-pok  to move forward for the circuit entity though.

dkeller9 commented:
Okay, I will push a version without the unstructured data.

dkeller9 commented on app/db/model.py:
```diff
@@ -76,24 +89,75 @@ class LegacyMixin:
+            uselist=True,
+            viewonly=True,
+            order_by="BrainRegion.brain_region_name"
+        )
```
Yes, specific_brain_regions is needed because there are artifacts that cannot be assigned to a specific brain region. We decided that they would always be children of the brain_region. IGiven the nature of data available in the literature, I think specific_brain_regions would be needed by most classes inheriting from ScientificArtifact. We can express it as an association_table I think. Is there an example in entitycore where this is done already for other features?

jdcourcol replied:
I am not sure I understand. What is the difference between brain_region and specific_brain_regions ? and why "specific" btw ?

GianlucaFicarelli replied:
> Given the nature of data available in the literature, I think specific_brain_regions would be needed by most classes inheriting from ScientificArtifact.

In that case, cannot we define only `specific_brain_regions` (and call it `brain_regions`), and remove `brain_region`?

An example of many-to-many association can be the one defined MTypesMixin, that uses the tables `mtype_classification` (the association table) and `mtype_class` (the table of unique mtypes).
Side note, mtype_class doesn't allow to specify different schemes of mtypes yet, but it can be changed in the future.

jdcourcol replied:
brain_regions is used in explore to filter. We are adding ScientificArtefact for circuit CNS use case. I would postpone that modification after CNS.

dkeller9 replied:
Many external databases and prior literature use atlas annotation schemes that do not map directly to ours. It is sometimes possible to assign a single coarse brain region but this does not capture the exact location of data. For that, we need specific_brain_regions which is a set of regions in our data that best reflects the original region assignment. I would be more than happy to remove "brain_region" entirely and just use a list of brain regions, but it is my understanding that some filtering operations are based on the assumption of one brain region. So keeping a list of specific_brain_regions and a single brain_region could allow compatability to be maintained.

dkeller9 commented on app/db/model.py:
```diff
@@ -102,88 +166,198 @@ class Strain(Identifiable):
+        "inherit_condition": id == Entity.id,
+    }
+
+    @staticmethod
+    def is_scientific_artifact(entity_type: EntityType) -> bool:
+        return entity_type in [
+            EntityType.cell_morphology,
+            EntityType.electrical_cell_recording,
+        ]
```
It is just a sanity check because there were potential issues in deployment with cell_morphology.

jdcourcol replied:
so let's remove it

dkeller9 replied:
removed

dkeller9 commented on app/db/model.py:
```diff
@@ -76,24 +89,75 @@ class LegacyMixin:
+    @classmethod
+    def __table_args__(cls):
+        # Define the index only for classes that own the description_vector column's table
+        if cls.__name__ in (
+            "ScientificArtifact",
+            "Subject",
+            "EModel",
+            "MEModel",
+            "Mesh",
+            "SingleNeuronSynaptome",
+            "SingleNeuronSimulation",
+            "SingleNeuronSynaptomeSimulation",
+            "ExperimentalNeuronDensity",
+            "ExperimentalBoutonDensity",
+            "ExperimentalSynapsesPerConnection",
+            "AnalysisSoftwareSourceCode"
+        ):
```
I had some odd issues during deployment and so added this check to be sure.

jdcourcol replied:
I would remove it then.

dkeller9 replied:
Okay, it is removed

dkeller9 commented on app/db/model.py:
```diff
@@ -38,16 +46,21 @@
     Sex,
     SingleNeuronSimulationStatus,
     ValidationStatus,
+    MorphologyType,
```
Yes, it is in a separate PR to be considered after all issues with ScientificArtifact are resolved. I already ran ScientificArtifact with the updated Morphology Schema, but the Morphology Schema is something that we cannot easily swap out yet because there are so many morphologies already in use.

dkeller9 replied:
We can remove MorphologyType before merging.

dkeller9 replied:
MorphologyType is removed since this PR is just ScientificArtifact.

eleftherioszisis commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,54 @@
+
+from app.filters.common import LocationFilterMixin
+
+class ScientificArtifactBase(LocationFilterMixin, BaseModel):
```
Schemas and filters are completely different things and cannot be mixed. Filters define how the user can query the read_many endpoint, whereas schemas determine the responses returned by the read/create endpoints.

dkeller9 replied:
Okay, I removed the filter dependency.

danielaegassan commented on app/schemas/scientific_artifact.py:
```diff
@@ -0,0 +1,55 @@
+    authorized_project_id: UUID
+    authorized_public: bool = False
+    createdBy_id: Optional[UUID] = None
+    updatedBy_id: Optional[UUID] = None
```
Why is this not a child of Entity instead where most of these fields are? 

Isn't CreatedBy mandatory since it's the person that registers the entity in the database?

Does PublishedIn allow more than one entry?  I think we need that.

dkeller9 replied:
Fixed in latest commit.

jdcourcol replied:
>Great, however, I'm still confused as to why we specify attributes in class ScientificArtifactBase that are already in entity?

yes this should be fixed.

jdcourcol replied:
>Also the issue with some of them being optional that I believe should be mandatory e.g., CreatedBy

I think @eleftherioszisis made a PR for that and that should not be optional

danielaegassan replied:
The thing is that if this is just inherited from Entity, then they will  be done consistently across all entities no? I don't get why we have to add them again.
